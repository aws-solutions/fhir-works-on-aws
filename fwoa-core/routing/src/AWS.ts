/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import './offlineEnvVariables';
import AWS from 'aws-sdk';
import AWSXRay from 'aws-xray-sdk';

const AWSWithXray = AWSXRay.captureAWS(AWS);

AWS.config.update({
  customUserAgent: process.env.CUSTOM_USER_AGENT
});

const { IS_OFFLINE } = process.env;
if (IS_OFFLINE === 'true') {
  AWS.config.update({
    region: process.env.AWS_REGION || 'us-west-2',
    accessKeyId: process.env.ACCESS_KEY,
    secretAccessKey: process.env.SECRET_KEY
  });
}

export default AWSWithXray;
