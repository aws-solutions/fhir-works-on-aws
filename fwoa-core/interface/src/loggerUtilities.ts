import AWS from 'aws-sdk';
import _ from 'lodash';

export async function encryptKMS(plaintext: string, keyId: string): Promise<string> {
  if (!plaintext) throw Error('Invalid input');
  const kms = new AWS.KMS();
  const params = {
    KeyId: keyId,
    Plaintext: plaintext
  };
  const encryptRes = await kms.encrypt(params).promise();
  const encryptedstring = encryptRes.CiphertextBlob;
  return Buffer.from(encryptedstring as Buffer).toString('base64');
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function encryptSelectedField(info: any): Promise<string> {
  let loggingMessage = { ...info.message };
  const field = info.meta.encryptedField;
  if (typeof field !== 'string' || !field.length) {
    throw new Error('Invalid field input to encrypt');
  }
  if (!_.has(loggingMessage, field)) {
    throw new Error('field does not exist in loggingMessage, cannot encrypt');
  }
  const fieldsContentsStringToEncrypt = JSON.stringify(_.get(loggingMessage, field), null, ' ');
  const encryptedFieldsContentsString = await encryptKMS(
    fieldsContentsStringToEncrypt,
    process.env.LOGGING_MIDDLEWARE_KMS_KEY as string
  );
  loggingMessage = _.omit(loggingMessage, field);
  _.set(loggingMessage, 'logMetadata.encryptedPayLoad', encryptedFieldsContentsString);
  return loggingMessage;
}

// Use console here so request ID and log level can be automatically attached in CloudWatch log
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function runLoggerLevel(info: any, msg: Array<{}>): void {
  /* eslint-disable no-console */
  switch (info[Symbol.for('level')]) {
    case 'debug':
      console.debug(...msg);
      break;
    case 'info':
      console.info(...msg);
      break;
    case 'warn':
      console.warn(...msg);
      break;
    case 'error':
      console.error(...msg);
      break;
    default:
      console.log(...msg);
      break;
  }
  /* eslint-enable no-console */
}
