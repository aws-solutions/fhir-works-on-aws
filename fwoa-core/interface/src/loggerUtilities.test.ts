/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import AWS from 'aws-sdk';
import * as AWSMock from 'aws-sdk-mock';
import inputLogger from './inputExampleEncryptSelectedField.json';
import { encryptKMS, encryptSelectedField } from './loggerUtilities';

process.env.LOGGING_MIDDLEWARE_KMS_KEY = 'fake-key-id';

describe('test logger utilities', () => {
  afterEach(() => {
    jest.clearAllMocks();
    AWSMock.restore('KMS');
  });
  describe('encryptSelectedField', () => {
    test('test happy case"', async () => {
      //BUILD
      const exampleMessage = inputLogger;
      const encryptRes: string = 'FakeEncryptedString';
      AWSMock.setSDKInstance(AWS);

      // eslint-disable-next-line @typescript-eslint/ban-types
      AWSMock.mock('KMS', 'encrypt', (params: { KeyId: string; Plaintext: string }, callback: Function) => {
        callback(null, {
          CiphertextBlob: Buffer.from(encryptRes),
          KeyId: '1233424123312',
          EncryptionAlgorithm: 'SYMMETRIC_DEFAULT'
        });
      });
      const info = {
        meta: { encryptedField: 'logMetadata.payLoadToEncrypt', component: 'routing' },
        message: exampleMessage
      };

      //OPERATION
      const result = await encryptSelectedField(info);

      //CHECK
      expect(result).toMatchSnapshot();
    });
    describe('invalid inputs', () => {
      test('test error: empty field name to encrypt', async () => {
        //BUILD
        const exampleMessage = inputLogger;
        const info = {
          meta: { encryptedField: '', componenet: 'routing' },
          message: exampleMessage
        };

        //CHECK
        await expect(encryptSelectedField(info)).rejects.toThrow('Invalid field input to encrypt');
      });
      test('test error: Invalid data type for field input', async () => {
        const exampleMessage = inputLogger;
        const info = {
          meta: { encryptedField: 2342, componenet: 'routing' },
          message: exampleMessage
        };
        await expect(encryptSelectedField(info)).rejects.toThrow('Invalid field input to encrypt');
      });
      test('test error: Invalid field name to encrypt', async () => {
        //BUILD
        const exampleMessage = inputLogger;
        const info = {
          meta: { encryptedField: 'asdas', componenet: 'routing' },
          message: exampleMessage
        };

        //CHECK
        await expect(encryptSelectedField(info)).rejects.toThrow(
          'field does not exist in loggingMessage, cannot encrypt'
        );
      });
    });
  });
  describe('encryptKMS', () => {
    test('Successfully encrypt string', async () => {
      // BUILD
      const encryptRes: string = 'FakeEncryptedString';
      AWSMock.setSDKInstance(AWS);
      //eslint-disable-next-line @typescript-eslint/ban-types
      AWSMock.mock('KMS', 'encrypt', (params: { KeyId: string; Plaintext: string }, callback: Function) => {
        callback(null, {
          CiphertextBlob: Buffer.from(encryptRes),
          KeyId: '1233424123312',
          EncryptionAlgorithm: 'SYMMETRIC_DEFAULT'
        });
      });

      // OPERATE
      const result = await encryptKMS('example', '1233424123312');

      // CHECK
      expect(result).toEqual(Buffer.from(encryptRes).toString('base64'));
    });
    test('Input string is empty', async () => {
      // BUILD
      const encryptRes: string = 'FakeEncryptedString';
      AWSMock.setSDKInstance(AWS);
      //eslint-disable-next-line @typescript-eslint/ban-types
      AWSMock.mock('KMS', 'encrypt', (params: { KeyId: string; Plaintext: string }, callback: Function) => {
        if (params.Plaintext) {
          callback(null, {
            CiphertextBlob: encryptRes,
            KeyId: '12314324324',
            EncryptionAlgorithm: 'SYMMETRIC_DEFAULT'
          });
        }
        callback(new Error('Invalid input'));
      });

      // OPERATE & CHECK
      await expect(encryptKMS('', '123456789012')).rejects.toThrowError('Invalid input');
    });
  });
});
