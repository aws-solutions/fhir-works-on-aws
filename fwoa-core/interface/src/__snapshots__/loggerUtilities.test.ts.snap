// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test logger utilities encryptSelectedField test happy case" 1`] = `
Object {
  "how": Object {
    "apiGateway": Object {
      "context": Object {
        "awsRequestId": "11111111-1111-1111-1111-111111111111",
      },
    },
    "userIdentity": Object {
      "jti": "AT.FakeJtiValue",
    },
  },
  "logMetadata": Object {
    "category": "Audit Event",
    "encryptedFields": Array [
      "who.userIdentity.sub",
      "who.userIdentity.fhirUser",
      "what.apiGateway.event.queryStringParameters",
      "what.requestContext.path",
      "what.apiGateway.event.pathParameters.proxy",
      "where.requestContext.sourceIp",
      "responseOther.userIdentity.launch-response-patient",
    ],
    "encryptedPayLoad": "RmFrZUVuY3J5cHRlZFN0cmluZw==",
    "timestamp": "2023-02-01T17:23:54.905Z",
    "uid": "11111111-1111-1111-1111-111111111111",
  },
  "requestOther": Object {
    "requestContext": Object {
      "stage": "dev",
    },
  },
  "responseOther": Object {
    "userIdentity": Object {
      "aud": "https://example.execute-api.us-east-1.amazonaws.com/dev",
      "auth-time": "2023-02-01T17:14:05.000Z",
      "exp": "2023-02-01T18:14:28.000Z",
      "iat": "2023-02-01T17:14:28.000Z",
      "iss": "https://dev-00000000.okta.com/oauth2/FakeIssuer",
      "launch-response-patient": "https://example.execute-api.us-east-1.amazonaws.com/dev/Patient/encrypted",
      "scp": Array [
        "FakeScope1",
        "FakeScope2",
        "FakeScope3",
      ],
    },
  },
  "what": Object {
    "apiGateway": Object {
      "event": Object {
        "httpMethod": "GET",
        "pathParameters": Object {
          "proxy": "Patient",
        },
        "queryStringParameters": "encrypted",
      },
    },
    "requestContext": Object {
      "httpMethod": "GET",
      "path": "/dev/Patient",
    },
    "userIdentity": Object {
      "scopes": Array [
        "FakeScope1",
        "FakeScope2",
        "FakeScope3",
      ],
      "usableScopes": Array [
        "FakeScope2",
        "FakeScope3",
      ],
    },
  },
  "when": Object {
    "requestContext": Object {
      "requestTimeEpoch": "2023-02-01T17:23:50.610Z",
    },
  },
  "where": Object {
    "User-agent": "FakeUserAgent",
    "apiGateway": Object {
      "Context": Object {
        "logGroupName": "/aws/lambda/smart-fhir-service-dev-fhirServer00000000-AAAAAAAAAAAA",
        "logStreamName": "2023/02/01/[$LATEST]FakeLogStreamId",
      },
    },
    "domainName": "example.execute-api.us-east-1.amazonaws.com",
    "requestContext": Object {
      "sourceIp": "encrypted",
    },
  },
  "who": Object {
    "apiKeyId": "x1siojg81g",
    "userIdentity": Object {
      "fhirUser": "https://example.execute-api.us-east-1.amazonaws.com/dev/Practitioner/encrypted",
      "sub": "encrypted",
    },
  },
}
`;
